### PR TITLE
2327: Property '{0}' is optional in type '{1}' but required in type '{2}'.

### PR DESCRIPTION
**Pull request by eddyw**
# Description

Screen Shot 2022-05-05 at 02 02 33

### CODE REVIEW

File: packages/engine/errors/2327.md [hunk 1]
```diff
@@ -0,0 +1,19 @@
+---
+original: "Property '{0}' is optional in type '{1}' but required in type '{2}'."
+excerpt: "Property '{0}' in type '{2}' cannot be 'undefined'."
```
Comment (mattpocock) [inline packages/engine/errors/2327.md:3]: I think this error can also show when you have:

`a: string | undefined`

being compared with:

`b?: string`.

That means that the excerpt above is slightly incorrect.

File: packages/engine/errors/2327.md [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+---
+
+You're assigning to a variable whose type `A` matches `B` but `B` has a property which can be `undefined` whereas `A` disallows it. For example:
```
Comment (mattpocock) [inline packages/engine/errors/2327.md:6]: I'd like somewhere in these explanations to have 1 and 2 to be displayed in full (this is where prettifying could come in handy)

**Review by mattpocock — CHANGES_REQUESTED** [review]
Some changes needed, see comments

File: packages/engine/errors/2327.md [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+```ts
+interface Foo {
+  {0}?: string
+}
+interface Bar {
+  {0}: string
+}
+const foo: Foo = { value: "" }
```
Comment (eddyw) [inline packages/engine/errors/2327.md:15]: 🤦‍♀️ 
```suggestion
const foo: Foo = { {0}: "" }
```

File: packages/engine/errors/2327.md (hunk 1 above)
Comment (eddyw) [reply packages/engine/errors/2327.md:3]: @mattpocock Hm, it's a different error what you're referring to:
```
Type 'string | undefined' is not assignable to type 'string'.
```
The error 2327 happens instead of the above when you have set `exactOptionalPropertyTypes` to true in compiler options:

```ts
interface Foo {
   value?: string
 }
declare function test(foo: Foo): void

test({value: undefined}) // <<< Error!!!
test({}) // <<< OK
```

It errors with `exactOptionalPropertyTypes: true` because `?` has a different meaning when that option is enabled, it means `value` can only be `string` or the property `value` must not exist in the object at all.

With `exactOptionalPropertyTypes: true`, you'd type it like this to achieve the behavior without this option enabled:

```ts
interface Foo {
   value?: string | undefined // <<< '?' but also ' | undefined'
 }
declare function test(foo: Foo): void

test({value: undefined}) // <<< OK!
test({}) // <<< OK
```

So to sum up, with this option `exactOptionalPropertyTypes: true` there is a new concept of "optional" and "required" properties. "optional" ones cannot be set "undefined" (unless explicitly set to `undefined` like in the last snippet)

Does this make sense? 😅 
Maybe I should also mention that the user is seeing this because they have `exactOptionalPropertyTypes: true`? 🤔 
After writing this comment, maybe I can do better at explaining this cases here 🤔 .. I'll revisit it later.

File: packages/engine/errors/2327.md [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+original: "Property '{0}' is optional in type '{1}' but required in type '{2}'."
+excerpt: "Property '{0}' in type '{2}' must exist."
+---
+
+Types `{1}` and `{2}` have the same property `{0}` but the property `{0}` is prefixed with `?` in type `{1}` which means it may not exist.
+
+For example, the interface `{1}` declares that there is a property `{0}` which can be a `string` or it should not exist in the object:
+
+```ts
+interface {1} {
```
Comment (mattpocock) [inline packages/engine/errors/2327.md:11]: I think that `{1}` can be represented by things other than a name - i.e. an object literal.

Comment (eddyw) [reply packages/engine/errors/2327.md:11]: @mattpocock hm not in this case. The error is explicit in that it'd include an interface name:
> (...) the **interface `{1}`** declares (...)

**Review by mattpocock — APPROVED** [review]
Makes sense, apologies for not getting to this sooner.
